### PR TITLE
Add a sample for the Cache command

### LINKED ISSUE #2399: Cache command doesn't have a sample
The `Cache` command doesn't have a sample. For 1.0, we need to have a sample, with links from the catalog xml docs.

### PR DESCRIPTION
**Pull request by rogancarr**
This PR adds a sample for the Cache command and updates the xml docs for the Cache catalog entry.

Fixes #2399

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            var enumerable = mlContext.CreateEnumerable<SamplesUtils.DatasetUtils.FloatLabelFloatFeatureVectorSample>(data, reuseRowObject: true);
+            foreach (var row in enumerable)
+            {
+                lines++;
+                averageOfColumn0 += row.Features[0];
+            }
+            watch.Stop();
+            averageOfColumn0 /= lines;
+            var elapsedSeconds = watch.ElapsedMilliseconds / 1000.0;
+            Console.WriteLine($"Lines={lines}, averageOfColumn0={averageOfColumn0:0.00} and took {elapsedSeconds} seconds.");
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: I suggest extracting this as a function, which is repeated 3 times. Repeating the same code will make the example look more complicated than what it actually is.

Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: Done.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        private static (int lines, double columnAverage, double elapsedSeconds) TimeToScanIDataView(MLContext mlContext, IDataView data)
+        {
+            int lines = 0;
+            double columnAverage = 0.0;
+            var enumerable = mlContext.CreateEnumerable<HousingRegression>(data, reuseRowObject: true);
+            var watch = System.Diagnostics.Stopwatch.StartNew();
+            foreach (var row in enumerable)
+            {
+                lines++;
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: >lines [](start = 16, length = 5)

can use the Count field in the enumerable, instead.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                columnAverage += row.MedianHomeValue + row.CrimesPerCapita + row.PercentResidental + row.PercentNonRetail + row.CharlesRiver 
+                    + row.NitricOxides + row.RoomsPerDwelling + row.PercentPre40s + row.EmploymentDistance 
+                    + row.HighwayDistance + row.TaxRate + row.TeacherRatio;
+            }
+            watch.Stop();
+            columnAverage /= lines;
+            var elapsedSeconds = watch.ElapsedMilliseconds / 1000.0;
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: > var elapsedSeconds = watch.ElapsedMilliseconds / 1000.0; [](start = 11, length = 57)

i think you can get the Elapsed time as a TimeSpan, and that will do the seconds conversion. 

TimeSpan ts = watch.Elapsed;
ts.Seconds.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            public float PercentNonRetail { get; set; }
+            public float CharlesRiver { get; set; }
+            public float NitricOxides { get; set; }
+            public float RoomsPerDwelling { get; set; }
+            public float PercentPre40s { get; set; }
+            public float EmploymentDistance { get; set; }
+            public float HighwayDistance { get; set; }
+            public float TaxRate { get; set; }
+            public float TeacherRatio { get; set; }
+        }
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: we've been trying to keep the sample support code here:

\Microsoft.ML.SamplesUtils\SamplesDatasetUtils.cs

so the samples display shorter, and more to the point. Maybe we can have this class there too.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 5]
```diff
@@ -26,13 +26,15 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
-        /// Caching happens per-column. A column is only cached when it is first accessed.
-        /// In addition, <paramref name="columnsToPrefetch"/> are considered 'always needed', so all of them
-        /// will be cached whenever any data is requested.
         /// </summary>
+        /// <remarks>
+        /// Caching happens per-column. A column is only cached when it is first accessed.
+        /// In addition, <paramref name="columnsToPrefetch"/> are considered 'always needed', so these columns
+        /// will be cached the first time any data is requested.
+        /// </remarks>
         /// <param name="input">The data view to cache.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >data view [](start = 36, length = 9)

<see cref="IDataView" />

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 6]
```diff
@@ -26,13 +26,15 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
-        /// <param name="columnsToPrefetch">The columns that must be cached whenever anything is cached. Empty array or null
-        /// is acceptable, it means that all columns are only cached at the first access.</param>
+        /// <param name="columnsToPrefetch">The columns that must be cached whenever anything is cached. An empty array or null
+        /// value means that columns are cached upon their first access.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >    /// valu [](start = 4, length = 12)

shall this link to the sample.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 2 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: I want to roll through exactly once and touch every field, so I'm doing a manual loop and tossing a counter in there. Count will iterate once just for itself and skew my timing calculations :(

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            var data = DatasetUtils.LoadHousingRegressionDataset(mlContext);
+
+            // Time how long it takes to page through the records if we don't cache.
+            (int lines, double columnAverage, double elapsedSeconds) = TimeToScanIDataView(mlContext, data);
+            Console.WriteLine($"Lines={lines}, averageOfColumn0={columnAverage:0.00} and took {elapsedSeconds} seconds.");
+            // Expected output (time is approximate):
+            // Lines=506, averageOfColumn0=564.17 and took 0.314 seconds.
+
+            // Now create a cached view of the data.
+            var cachedData = mlContext.Data.Cache(data);
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: I think this is a great idea. I'd love to write practical samples and use cases for ML.NET once we have these simple API samples completed.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 6 above)
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Yes, good catch!

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 5 above)
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Used the same text as the other filters.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 4 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: Done!

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (justinormont) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: Yes, and I think the user should always be calling `WantCaching` as the check is free (assuming it's even public still). Hence we should teach this method.

File: docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs [hunk 8]
```diff
@@ -1,47 +1,21 @@
...
-                hasHeader: true
-            );
             
-            // Read the data
-            var data = reader.Read(dataFile);
+            // Read the Housing regression dataset
+            var data = DatasetUtils.LoadHousingRegressionDataset(mlContext);
 
-            // Step 2: Pipeline
+            // Step 1: Pipeline
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/GeneralizedAdditiveModels.cs]: >Step 1 [](start = 15, length = 6)

are we keeping the 'step' or dropping them? you removed it from fcc sample.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 2 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: or you could initialize lines to enumerable.Count() before the stopwatch

---
In reply to: [253984113](https://github.com/dotnet/machinelearning/pull/2406#discussion_r253984113) [](ancestors = 253984113)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: Can you file this as a separate issue?

---
In reply to: [254021643](https://github.com/dotnet/machinelearning/pull/2406#discussion_r254021643) [](ancestors = 254021643)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 4 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: Done!

---
In reply to: [253741427](https://github.com/dotnet/machinelearning/pull/2406#discussion_r253741427) [](ancestors = 253741427)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs (hunk 3 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/Cache.cs]: Nice.

---
In reply to: [253741237](https://github.com/dotnet/machinelearning/pull/2406#discussion_r253741237) [](ancestors = 253741237)

**Pull request merged**